producer/awsS3.go

```diff
@@ -0,0 +1,281 @@
+// Copyright 2015-2017 trivago GmbH
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package producer
+
+import (
+	"fmt"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/trivago/gollum/core"
+	"github.com/trivago/gollum/core/components"
+	"github.com/trivago/gollum/producer/awsS3"
+	"github.com/trivago/tgo/tmath"
+	"path/filepath"
+	"strings"
+	"sync"
+	"time"
+)
+
+// AwsS3 producer plugin
+//
+// Configuration example
+//
+// myProducer:
+//   Type: producer.AwsS3
+//   Credentials:
+//     Type: shared
+//     File: /Users/$MYUSER/.aws/credentials
+//     Profile: default
+//   Region: eu-west-1
+//   Bucket: my-s3-bucket/subfolder/
+//   Streams:
+//     - "*"
+//   File: gollum_*.log
+//   Batch:
+//     TimeoutSec: 5
+//     MaxCount: 1000
+//     FlushCount: 500
+//     FlushTimeoutSec: 0
+//   Rotation:
+//     TimestampFormat: 2006-01-02_15
+//     TimeoutMin: 1440
+//     SizeMB: 1024
+// 	   ZeroPadding: 0
+// 	   At: 13:05
+//   Modulators:
+//     - format.Envelope:
+//       Postfix: "\n"
+//
+type AwsS3 struct {
+	core.DirectProducer `gollumdoc:"embed_type"`
+
+	// Rotate is public to make Pruner.Configure() callable (bug in treflect package)
+	Rotate         components.RotateConfig   `gollumdoc:"embed_type"`
+	AwsMultiClient components.AwsMultiClient `gollumdoc:"embed_type"`
+
+	// configuration
+	batchTimeout      time.Duration `config:"Batch/TimeoutSec" default:"5" metric:"sec"`
+	batchMaxCount     int           `config:"Batch/MaxCount" default:"1000"`
+	batchFlushCount   int           `config:"Batch/FlushCount" default:"500"`
+	batchFlushTimeout time.Duration `config:"Batch/FlushTimeoutSec" default:"0" metric:"sec"`
+	bucket            string        `config:"Bucket" default:""`
```

Comment:
Take this block and move it into a separate type. Make this type a factory for batchedFiles.
```go
type BatchedFileBuilder struct {
	batchTimeout      time.Duration `config:"Batch/TimeoutSec" default:"5" metric:"sec"`
 	batchMaxCount     int           `config:"Batch/MaxCount" default:"1000"`
 	batchFlushCount   int           `config:"Batch/FlushCount" default:"500"`
 	batchFlushTimeout time.Duration `config:"Batch/FlushTimeoutSec" default:"0" metric:"sec"`
}
```

... continued below